File: python/test/base/multi_document_assembler_test.py
#  Copyright 2017-2022 John Snow Labs
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
import unittest

import pytest

from sparknlp.base import *
from sparknlp.annotator import *
from test.util import SparkSessionForTest


@pytest.mark.fast
class MultiDocumentAssemblerTest(unittest.TestCase):

    def setUp(self):
        self.spark = SparkSessionForTest.spark
        self.input1 = "This is the first input"
        self.input2 = "This is the second input"
        self.twoInputDataSet = self.spark.createDataFrame([[self.input1, self.input2]]) \
            .toDF("input1", "input2")

    def runTest(self):

        multi_document_assembler = MultiDocumentAssembler() \
            .setInputCols("input1", "input2") \
            .setOutputCols("output1", "output2")

        tokenizer = Tokenizer() \
            .setInputCols("output2") \
            .setOutputCol("token")

        pipeline = Pipeline().setStages([multi_document_assembler, tokenizer])
        result_df = pipeline.fit(self.twoInputDataSet).transform(self.twoInputDataSet)

        output1 = result_df.select("output1.result").collect()[0]
        output2 = result_df.select("output2.result").collect()[0]
        tokens = result_df.select("token.result").collect()[0]

        assert len(output1) > 0
        assert len(output2) > 0
        assert len(tokens) > 0


@pytest.mark.fast
class LightFullAnnotateMultiDocumentAssemblerTest(unittest.TestCase):

    def setUp(self):
        self.spark = SparkSessionForTest.spark
        self.input1 = "This is the first input"
        self.input2 = "This is the second input"
        self.twoInputDataSet = self.spark.createDataFrame([[self.input1, self.input2]]) \
            .toDF("input1", "input2")

    def runTest(self):

        multi_document_assembler = MultiDocumentAssembler() \
            .setInputCols("input1", "input2") \
            .setOutputCols("output1", "output2")

        tokenizer = Tokenizer() \
            .setInputCols("output2") \
            .setOutputCol("token")

        pipeline = Pipeline().setStages([multi_document_assembler, tokenizer])
        model = pipeline.fit(self.twoInputDataSet)
        light_pipeline = LightPipeline(model)

        actual_result = light_pipeline.fullAnnotate(self.input1, self.input2)
        expected_result = [{
            "output1": [Annotation("document", 0, len(self.input1) - 1, self.input1, {}, [])],
            "output2": [Annotation("document", 0, len(self.input2) - 1, self.input2, {}, [])],
            "token": [
                Annotation("token", 0, 3, "This", {"sentence": "0"}, []),
                Annotation("token", 5, 6, "is", {"sentence": "0"}, []),
                Annotation("token", 8, 10, "the", {"sentence": "0"}, []),
                Annotation("token", 12, 17, "second", {"sentence": "0"}, []),
                Annotation("token", 19, 23, "input", {"sentence": "0"}, [])
            ]
        }]

        assert actual_result == expected_result

        actual_result2 = light_pipeline.fullAnnotate([[self.input1, self.input2], [self.input1, self.input2]])
        expected_result2 = [{
            "output1": [Annotation("document", 0, len(self.input1) - 1, self.input1, {}, [])],
            "output2": [Annotation("document", 0, len(self.input2) - 1, self.input2, {}, [])],
            "token": [
                Annotation("token", 0, 3, "This", {"sentence": "0"}, []),
                Annotation("token", 5, 6, "is", {"sentence": "0"}, []),
                Annotation("token", 8, 10, "the", {"sentence": "0"}, []),
                Annotation("token", 12, 17, "second", {"sentence": "0"}, []),
                Annotation("token", 19, 23, "input", {"sentence": "0"}, [])
            ]
            },
            {
                "output1": [Annotation("document", 0, len(self.input1) - 1, self.input1, {}, [])],
                "output2": [Annotation("document", 0, len(self.input2) - 1, self.input2, {}, [])],
                "token": [
                    Annotation("token", 0, 3, "This", {"sentence": "0"}, []),
                    Annotation("token", 5, 6, "is", {"sentence": "0"}, []),
                    Annotation("token", 8, 10, "the", {"sentence": "0"}, []),
                    Annotation("token", 12, 17, "second", {"sentence": "0"}, []),
                    Annotation("token", 19, 23, "input", {"sentence": "0"}, [])
                ]
            }
        ]

        assert actual_result2 == expected_result2


@pytest.mark.fast
class LightAnnotateMultiDocumentAssemblerTest(unittest.TestCase):

    def setUp(self):
        self.spark = SparkSessionForTest.spark
        self.input1 = "This is the first input"
        self.input2 = "This is the second input"
        self.twoInputDataSet = self.spark.createDataFrame([[self.input1, self.input2]]) \
            .toDF("input1", "input2")

    def runTest(self):

        multi_document_assembler = MultiDocumentAssembler() \
            .setInputCols("input1", "input2") \
            .setOutputCols("output1", "output2")

        tokenizer = Tokenizer() \
            .setInputCols("output2") \
            .setOutputCol("token")

        pipeline = Pipeline().setStages([multi_document_assembler, tokenizer])
        model = pipeline.fit(self.twoInputDataSet)
        light_pipeline = LightPipeline(model)

        actual_result = light_pipeline.annotate(self.input1, self.input2)
        expected_result = {
            "output1": [self.input1],
            "output2": [self.input2],
            "token": ["This", "is", "the", "second", "input"]
        }

        assert actual_result == expected_result

        actual_result2 = light_pipeline.annotate([[self.input1, self.input2], [self.input1, self.input2]])
        expected_result2 = [{
            "output1": [self.input1],
            "output2": [self.input2],
            "token": ["This", "is", "the", "second", "input"]
            },
            {
            "output1": [self.input1],
            "output2": [self.input2],
            "token": ["This", "is", "the", "second", "input"]
            }
        ]

        assert actual_result2 == expected_result2

File: src/test/scala/com/johnsnowlabs/nlp/MultiDocumentAssemblerTest.scala
package com.johnsnowlabs.nlp

import com.johnsnowlabs.nlp.AnnotatorType.{DOCUMENT, TOKEN}
import com.johnsnowlabs.nlp.annotators.SparkSessionTest
import com.johnsnowlabs.tags.FastTest
import org.scalatest.flatspec.AnyFlatSpec

class MultiDocumentAssemblerTest extends AnyFlatSpec with SparkSessionTest {

  import spark.implicits._

  val input1 = "This is the first input"
  val input2 = "This is the second input"

  private val twoInputDataset = Seq((input1, input2)).toDS.toDF("input1", "input2")
  private val dataset = Seq(input1).toDS.toDF("text")

  "MultiDocumentAssembler with two input cols" should "transform and output two cols" taggedAs FastTest in {

    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("input1", "input2")
      .setOutputCols("output1", "output2")
    tokenizer.setInputCols("output2").setOutputCol("token")

    val multiDocumentPipeline = pipeline.setStages(Array(multiDocumentAssembler, tokenizer))

    val result = multiDocumentPipeline.fit(twoInputDataset).transform(twoInputDataset)

    val actualOutput1 = AssertAnnotations.getActualResult(result, "output1")
    val actualOutput2 = AssertAnnotations.getActualResult(result, "output2")
    val actualResultToken = AssertAnnotations.getActualResult(result, "token")

    val expectedOutput1 =
      Array(Seq(Annotation(DOCUMENT, 0, input1.length - 1, input1, Map("sentence" -> "0"))))
    val expectedOutput2 =
      Array(Seq(Annotation(DOCUMENT, 0, input2.length - 1, input2, Map("sentence" -> "0"))))
    val expectedResultToken = Array(
      Seq(
        Annotation(TOKEN, 0, 3, "This", Map("sentence" -> "0")),
        Annotation(TOKEN, 5, 6, "is", Map("sentence" -> "0")),
        Annotation(TOKEN, 8, 10, "the", Map("sentence" -> "0")),
        Annotation(TOKEN, 12, 17, "second", Map("sentence" -> "0")),
        Annotation(TOKEN, 19, 23, "input", Map("sentence" -> "0"))))

    AssertAnnotations.assertFields(actualOutput1, expectedOutput1)
    AssertAnnotations.assertFields(actualOutput2, expectedOutput2)
    AssertAnnotations.assertFields(actualResultToken, expectedResultToken)
  }

  it should "fullAnnotate with LightPipeline and output two cols" taggedAs FastTest in {
    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("input1", "input2")
      .setOutputCols("output1", "output2")

    tokenizer.setInputCols("output2").setOutputCol("token")

    val pipelineModel =
      pipeline.setStages(Array(multiDocumentAssembler, tokenizer)).fit(twoInputDataset)

    val lightPipeline = new LightPipeline(pipelineModel)
    val actualResult = lightPipeline.fullAnnotate(input1, input2)

    val expectedResult =
      Map(
        "output1" -> List(Annotation(DOCUMENT, 0, input1.length - 1, input1, Map())),
        "output2" -> List(Annotation(DOCUMENT, 0, input2.length - 1, input2, Map())),
        "token" -> List(
          Annotation(TOKEN, 0, 3, "This", Map("sentence" -> "0")),
          Annotation(TOKEN, 5, 6, "is", Map("sentence" -> "0")),
          Annotation(TOKEN, 8, 10, "the", Map("sentence" -> "0")),
          Annotation(TOKEN, 12, 17, "second", Map("sentence" -> "0")),
          Annotation(TOKEN, 19, 23, "input", Map("sentence" -> "0"))))

    assert(actualResult.keySet == expectedResult.keySet)
    AssertAnnotations.assertFields(expectedResult.values.toArray, actualResult.values.toArray)

  }

  it should "fullAnnotate a list of inputs with LightPipeline and output of two cols" taggedAs FastTest in {
    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("input1", "input2")
      .setOutputCols("output1", "output2")

    tokenizer.setInputCols("output2").setOutputCol("token")

    val pipelineModel =
      pipeline.setStages(Array(multiDocumentAssembler, tokenizer)).fit(twoInputDataset)

    val lightPipeline = new LightPipeline(pipelineModel)
    val actualResults =
      lightPipeline.fullAnnotate(Array(input1, input1), Array(input2, input2))

    val expectedResults = Array(
      Map(
        "output1" -> List(Annotation(DOCUMENT, 0, input1.length - 1, input1, Map())),
        "output2" -> List(Annotation(DOCUMENT, 0, input2.length - 1, input2, Map())),
        "token" -> List(
          Annotation(TOKEN, 0, 3, "This", Map("sentence" -> "0")),
          Annotation(TOKEN, 5, 6, "is", Map("sentence" -> "0")),
          Annotation(TOKEN, 8, 10, "the", Map("sentence" -> "0")),
          Annotation(TOKEN, 12, 17, "second", Map("sentence" -> "0")),
          Annotation(TOKEN, 19, 23, "input", Map("sentence" -> "0")))),
      Map(
        "output1" -> List(Annotation(DOCUMENT, 0, input1.length - 1, input1, Map())),
        "output2" -> List(Annotation(DOCUMENT, 0, input2.length - 1, input2, Map())),
        "token" -> List(
          Annotation(TOKEN, 0, 3, "This", Map("sentence" -> "0")),
          Annotation(TOKEN, 5, 6, "is", Map("sentence" -> "0")),
          Annotation(TOKEN, 8, 10, "the", Map("sentence" -> "0")),
          Annotation(TOKEN, 12, 17, "second", Map("sentence" -> "0")),
          Annotation(TOKEN, 19, 23, "input", Map("sentence" -> "0")))))

    actualResults.zipWithIndex.foreach { case (actualResult, index) =>
      val expectedResult = expectedResults(index)
      assert(actualResult.keySet == expectedResult.keySet)
      AssertAnnotations.assertFields(expectedResult.values.toArray, actualResult.values.toArray)
    }
  }

  it should "annotate with LightPipeline and output two cols" taggedAs FastTest in {
    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("input1", "input2")
      .setOutputCols("output1", "output2")

    tokenizer.setInputCols("output2").setOutputCol("token")

    val pipelineModel =
      pipeline.setStages(Array(multiDocumentAssembler, tokenizer)).fit(twoInputDataset)

    val lightPipeline = new LightPipeline(pipelineModel)
    val actualResult = lightPipeline.annotate(input1, input2)

    val expectedResult =
      Map(
        "output1" -> List(input1),
        "output2" -> List(input2),
        "token" -> List("This", "is", "the", "second", "input"))

    assert(actualResult.keySet == expectedResult.keySet)
    assert(actualResult.values.toList == expectedResult.values.toList)
  }

  it should "annotate a list of inputs with LightPipeline and output two cols" taggedAs FastTest in {
    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("input1", "input2")
      .setOutputCols("output1", "output2")

    tokenizer.setInputCols("output2").setOutputCol("token")

    val pipelineModel =
      pipeline.setStages(Array(multiDocumentAssembler, tokenizer)).fit(twoInputDataset)

    val lightPipeline = new LightPipeline(pipelineModel)
    val actualResults =
      lightPipeline.annotate(Array(input1, input1), Array(input2, input2))

    val expectedResults = Array(
      Map(
        "output1" -> List(input1),
        "output2" -> List(input2),
        "token" -> List("This", "is", "the", "second", "input")),
      Map(
        "output1" -> List(input1),
        "output2" -> List(input2),
        "token" -> List("This", "is", "the", "second", "input")))

    actualResults.zipWithIndex.foreach { case (actualResult, index) =>
      val expectedResult = expectedResults(index)
      assert(actualResult.keySet == expectedResult.keySet)
      assert(actualResult.values.toList == expectedResult.values.toList)
    }

  }

  "MultiDocumentAssembler with one column" should "transform and output one col" taggedAs FastTest in {

    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("text")
      .setOutputCols("output")
    tokenizer.setInputCols("output").setOutputCol("token")

    val multiDocumentPipeline = pipeline.setStages(Array(multiDocumentAssembler, tokenizer))

    val result = multiDocumentPipeline.fit(dataset).transform(dataset)
    val actualOutput = AssertAnnotations.getActualResult(result, "output")
    val actualResultToken = AssertAnnotations.getActualResult(result, "token")

    val expectedOutput =
      Array(Seq(Annotation(DOCUMENT, 0, input1.length - 1, input1, Map("sentence" -> "0"))))
    val expectedResultToken = Array(
      Seq(
        Annotation(TOKEN, 0, 3, "This", Map("sentence" -> "0")),
        Annotation(TOKEN, 5, 6, "is", Map("sentence" -> "0")),
        Annotation(TOKEN, 8, 10, "the", Map("sentence" -> "0")),
        Annotation(TOKEN, 12, 16, "first", Map("sentence" -> "0")),
        Annotation(TOKEN, 18, 22, "input", Map("sentence" -> "0"))))

    AssertAnnotations.assertFields(actualOutput, expectedOutput)
    AssertAnnotations.assertFields(actualResultToken, expectedResultToken)
  }

  it should "fullAnnotate with LightPipeline and output one col" taggedAs FastTest in {
    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("text")
      .setOutputCols("output")
    tokenizer.setInputCols("output").setOutputCol("token")

    val pipelineModel = pipeline.setStages(Array(multiDocumentAssembler, tokenizer)).fit(dataset)

    val lightPipeline = new LightPipeline(pipelineModel)
    val actualResult = lightPipeline.fullAnnotate(input1)

    val expectedResult =
      Map(
        "output" -> List(Annotation(DOCUMENT, 0, input1.length - 1, input1, Map())),
        "token" -> List(
          Annotation(TOKEN, 0, 3, "This", Map("sentence" -> "0")),
          Annotation(TOKEN, 5, 6, "is", Map("sentence" -> "0")),
          Annotation(TOKEN, 8, 10, "the", Map("sentence" -> "0")),
          Annotation(TOKEN, 12, 16, "first", Map("sentence" -> "0")),
          Annotation(TOKEN, 18, 22, "input", Map("sentence" -> "0"))))

    assert(actualResult.keySet == expectedResult.keySet)
    AssertAnnotations.assertFields(expectedResult.values.toArray, actualResult.values.toArray)
  }

  it should "annotate with LightPipeline and output one col" taggedAs FastTest in {
    val multiDocumentAssembler = new MultiDocumentAssembler()
      .setInputCols("text")
      .setOutputCols("output")

    tokenizer.setInputCols("output").setOutputCol("token")

    val pipelineModel = pipeline.setStages(Array(multiDocumentAssembler, tokenizer)).fit(dataset)

    val lightPipeline = new LightPipeline(pipelineModel)
    val actualResult = lightPipeline.annotate(input1)

    val expectedResult =
      Map("output" -> List(input1), "token" -> List("This", "is", "the", "first", "input"))

    assert(actualResult.keySet == expectedResult.keySet)
    assert(actualResult.values.toList == expectedResult.values.toList)
  }

}
